### PR TITLE
Rewrite rd_kw_type as a c++ class rd::KW

### PR DESCRIPTION
**Pull request by eivindjahren**

### CODE REVIEW

File: lib/resdata/rd_kw_pybind.cpp [hunk 1]
```diff
@@ -141,216 +144,343 @@ PYBIND11_MODULE(_kw, m) {
...
+        return from_cwrap<rd::KW>(self)->at<std::string>(index);
     });
-    m.def("_iset_char_ptr", [](py::handle self, int index, std::string value) {
-        rd_kw_iset_char_ptr(from_cwrap<rd_kw_type>(self), index, value.c_str());
-    });
-    m.def("_iget_string_ptr", [](py::handle self, int index) {
-        return rd_kw_iget_string_ptr(from_cwrap<rd_kw_type>(self), index);
+    m.def("_iset_char_ptr",
+          [](py::handle self, size_t index, std::string value) {
+              from_cwrap<rd::KW>(self)->set_string_array(index, value);
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply lib/resdata/rd_kw_pybind.cpp:216]: This is just preserving the behavior existing behavior which called `rd_kw_iset_char_ptr`

File: lib/resdata/rd_kw_pybind.cpp [hunk 2]
```diff
@@ -53,72 +56,72 @@ class KwIterator {
...
         const std::string fmt = rd::format_kw_element_fmt(8);
         std::vector<char> buf(8 + 4);
-        for (int i = 0; i < size; i++) {
-            size_t len = rd::format_kw_element_buf(
-                buf.data(), buf.size(), rd_kw_iget_char_ptr(kw, i), fmt);
+        for (size_t i = 0; i < size; i++) {
+            const std::string value = kw->at<std::string>(i);
+            size_t len = rd::format_kw_element_buf(buf.data(), buf.size(),
+                                                   value.c_str(), fmt);
             tokens[i] = py::str(buf.data(), len);
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply lib/resdata/rd_kw_pybind.cpp:114]: This was fixed by using
```c++
inline std::string format_kw_element_fmt(size_t width) {
    return " '%-" + std::to_string(width) + "." + std::to_string(width) + "s'";
}
```

File: lib/resdata/rd_kw.cpp [hunk 3]
```diff
@@ -1316,1131 +828,429 @@ rd_read_status_enum rd_kw_fread_header(rd_kw_type *rd_kw, ERT::FortIO &fortio) {
...
+        for (size_t index = offset; index < this->size(); index++)
+            if (!elm_equal_numeric(this, rd_kw2, index, abs_epsilon,
+                                   rel_epsilon))
+                return index;
+        return this->size();
+    } else {
+        auto vec1 = this->data().value();
+        auto vec2 = rd_kw2->data().value();
+        return std::visit(
+            [vec2, offset](const auto &v1) -> size_t {
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply lib/resdata/rd_kw.cpp]: Resolved by using references.

File: lib/resdata/rd_kw.cpp [hunk 4]
```diff
@@ -619,79 +296,50 @@ rd_kw_type *rd_kw_alloc_copy(const rd_kw_type *src) {
...
-    int src_index = index1;
+    size_t src_index = index1;
     /* 1: Determine size of the sliced copy. */
-    int new_size = 0;
+    size_t new_size = 0;
     while (src_index < index2) {
         new_size++;
         src_index += stride;
     }
     if (new_size > 0) {
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply lib/resdata/rd_kw.cpp]: Resolved by simply dropping the conditional.

File: lib/include/resdata/FortIO.hpp [hunk 5]
```diff
@@ -46,8 +50,6 @@ class FortIO {
     FortIO() = delete;
     FortIO(const std::string &filename, std::ios_base::openmode mode,
            bool fmt_file = false, bool endian_flip_header = RD_ENDIAN_FLIP);
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply lib/include/resdata/FortIO.hpp:52]: This is intended

File: lib/include/resdata/rd_kw.hpp [hunk 6]
```diff
@@ -1,56 +1,309 @@
...
+            throw std::invalid_argument(
+                fmt::format("rd_kw size was negative: {}", size));
+        this->m_size = static_cast<size_t>(size);
+        zero_init_data();
+    }
+
+    template <typename T>
+    KW(const std::string &header, const std::vector<T> &data)
+        : m_size(data.size()), m_data_type(datatype<T>::tag),
+          m_header(strip_header(header)) {
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply lib/include/resdata/rd_kw.hpp:111]: This is intended as we want to avoid calling this constructor with std::string.

File: lib/include/resdata/rd_type.hpp [hunk 7]
```diff
@@ -108,6 +108,24 @@ template <> struct iotype<int> {
     static constexpr rd_type_enum tag = RD_INT_TYPE;
 };
 
+template <typename T> struct datatype {
+    static_assert(always_false<T>,
+                  "Type cannot be read element-wise from a keyword data "
+                  "section; expected float, double or int");
+};
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply lib/include/resdata/rd_type.hpp:115]: This is intended as we want to avoid calling those specializations with std::string.

File: lib/resdata/rd_smspec.cpp [hunk 8]
```diff
@@ -313,70 +313,42 @@ static rd_data_type get_wgnames_type(const rd_smspec_type *smspec) {
...
-        for (size_t i = 0; i < SUMMARY_RESTART_SIZE; i++) {
-            if (offset < restart_case_len)
-                rd_kw_iset_string8(restart_kw.get(), i,
-                                   &smspec->restart_case[offset]);
-            offset += RD_STRING8_LENGTH;
-        }
-    }
-    rd_kw_fwrite(restart_kw.get(), fortio);
+    rd::KW restart_kw{RESTART_KW, SUMMARY_RESTART_SIZE, RD_CHAR};
+    restart_kw.set_string_array(0, smspec->restart_case);
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply lib/resdata/rd_smspec.cpp:328]: ctype_size() is actually 9 for RD_CHAR so this uses 8-characters on disk.

File: lib/tests/test_rd_kw.cpp [hunk 9]
```diff
@@ -14,277 +15,158 @@
...
 }
 
-TEST_CASE("rd_kw_iget_stripped_string handles width edge cases", "[rd_kw]") {
+TEST_CASE("rd::strip_spaces handles width edge cases", "[rd_kw]") {
     SECTION("RD_CHAR values can fill the full field width") {
-        auto char_kw = make_rd_kw("KW", 2, RD_CHAR);
-
-        rd_kw_iset_char_ptr(char_kw.get(), 0, "FOPRTEST");
-        rd_kw_iset_char_ptr(char_kw.get(), 1, "BPR");
+        rd::KW char_kw{"KW", {"FOPRTEST", "BPR"}};
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply lib/tests/test_rd_kw.cpp:99]: ctype_size() is actually 9 for CHAR.

File: lib/include/resdata/rd_file_view.hpp [hunk 10]
```diff
@@ -185,8 +186,10 @@ class FileView {
 
     bool has_report_step(int report_step) {
         return find_block(SEQNUM_KW,
-                          [&](const rd_kw_type *seqnum_kw) {
-                              return rd_kw_data_equal(seqnum_kw, &report_step);
+                          [&](const rd::KW *seqnum_kw) {
+                              return seqnum_kw->size() > 0 &&
+                                     rd_type_is_int(seqnum_kw->data_type()) &&
+                                     seqnum_kw->at<int>(0) == report_step;
                           })
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply lib/include/resdata/rd_file_view.hpp:193]: This is too strict, we want to allow for leniency here.

File: lib/resdata/rd_file_view.cpp [hunk 11]
```diff
@@ -335,8 +335,10 @@ FileView::restart_view_from_seqnum_index(size_t index) {
 }
 std::shared_ptr<FileView>
 FileView::restart_view_from_report_step(int report_step) {
-    auto block = find_block(SEQNUM_KW, [&](const rd_kw_type *seqnum_kw) {
-        return rd_kw_data_equal(seqnum_kw, &report_step);
+    auto block = find_block(SEQNUM_KW, [&](const rd::KW *seqnum_kw) {
+        return seqnum_kw->size() > 0 &&
+               rd_type_is_int(seqnum_kw->data_type()) &&
+               seqnum_kw->at<int>(0) == report_step;
     });
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply lib/resdata/rd_file_view.cpp:342]: This is too strict, we want to allow for more than one element.
